Fixes #149 - upgrade medical_prescription_state_verify to v10

kellylougheed opened the pull request:
Upgrades medical_prescription_state_verify to v10 by eliminating the states and using only the stages.

kellylougheed committed:
Fixes #149 - Upgrade medical_prescription_state_verify to v10

woodbrettm commented on medical_prescription_state_verify/__manifest__.py:
```diff
@@ -8,14 +8,14 @@
     'author': "LasLabs, Odoo Medical Team, Odoo Community Association (OCA)",
```
Odoo Medical Team is an old entity, remove.

woodbrettm commented on medical_prescription_state_verify/__manifest__.py:
```diff
@@ -8,14 +8,14 @@
         'data/medical_prescription_order_state_data.xml',
     ],
-    'installable': False,
+    'installable': True,
```
Structure should follow like so:

```
{
    "name": "Odoo Medical",
    "summary": "Extends Odoo with medical patients, doctors, and centers.",
    "version": "10.0.1.0.0",
    "category": "Medical",
    "website": "https://laslabs.com",
    "author": "LasLabs, Odoo Community Association (OCA)",
    "license": "AGPL-3",
    "application": False,
    "installable": True,
    "depends": [
        "base",
        "product",
    ],
    "data": [
        "data/ir_sequence_data.xml",
        "security/medical_security.xml",
        "security/ir.model.access.csv",
        "views/medical_patient_view.xml",
        "views/res_partner_view.xml",
        "views/medical_menu.xml",
    ],
    "demo": [
        "demo/medical_patient_demo.xml",
    ],
}
```

woodbrettm commented on medical_prescription_state_verify/models/medical_prescription_order.py:
```diff
@@ -2,7 +2,7 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import fields, models, exceptions, api, _
+from odoo import api, exceptions, models, _
```
I learned from some of Ted's reviews of my code to put the `_` first, like so: `_, api, models`.
It's also usually more common to put `from odoo.exceptions import ValidationError` on a separate line, so you can do `raise ValidationError` in the file instead of `raise exceptions.ValidationError`.

woodbrettm commented on medical_prescription_state_verify/models/medical_prescription_order.py:
```diff
@@ -2,7 +2,7 @@
 # © 2016 LasLabs Inc.
```
Fix copyrights to `# Copyright 2016-2017 Laslabs Inc.`

woodbrettm reviewed (changes requested):
Thanks @kellylougheed 

Couple things to change. 
Also, the test coverage is around 80%. Could you raise that up to around 90-100%? Helps with catching bugs better. 

* If you want, a good exercise might be to replace all the tests with using demo data. medical_prescription in laslabs/vertical-medical/release/10.0 is a good example. 

I think the licenses need to be all LGPL-3.0 as well.

lasley commented:
> I think the licenses need to be all LGPL-3.0 

This change hasn't been made on 10 yet, I'm still splitting the modules up in #123

kellylougheed committed:
Changes based on feedback, add test

kellylougheed commented:
Made changes @bmw95! I added one test, but if you can think of anything else to test that hasn't been covered, I'm open to suggestions!

lasley commented:
Coverage is now 100%

woodbrettm commented on medical_prescription_state_verify/data/medical_prescription_order_state_data.xml:
```diff
@@ -1,72 +1,27 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
```xml
<?xml version="1.0" encoding="utf-8"?>
<!-- Copyright 2016-2017 LasLabs Inc.
     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
```

woodbrettm commented on medical_prescription_state_verify/__manifest__.py:
```diff
@@ -1,21 +1,21 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
 {
     'name': 'Medical Prescription Order State Verification',
```
All single quotes in the file should be double. (e.g. 'category': 'Medical'  => "category": "Medical")

woodbrettm reviewed (approved):
LGTM! Just 2 minor suggestions.

woodbrettm commented:
Just noticed a few other things:
1. README needs to be updated. Should be similar to [this](https://raw.githubusercontent.com/LasLabs/vertical-medical/release/10.0/medical_prescription/README.rst).
2. Remove the `from . import tests` from the main `__init__.py` file
3. Missing the /static/description/icon.png file (oca icon)

kellylougheed committed:
Changes in response to feedback
*Move Exception states to main stage module and reorder stages
*Add is_verified field to base_kanban_stage
*Change validation check from name 'Verified' to is_verified field
*Change _ALLOWED_CHANGE_STATES to be stage names

kellylougheed commented:
Changes made @bmw95!

woodbrettm reviewed (approved):
Thanks @kellylougheed LGTM!

lasley commented on medical_prescription_state_verify/data/medical_prescription_order_state_data.xml:
```diff
@@ -1,72 +1,23 @@
-        
-    </data>
+    <record id="prescription_order_state_exception" model="base.kanban.stage">
+        <field name="name">Exception</field>
```
Rx and Rx line states are still kind of up in the air. Let's add Exception to the main stage module instead

lasley commented on medical_prescription_state_verify/data/medical_prescription_order_state_data.xml:
```diff
@@ -1,72 +1,23 @@
+    </record>
+
+    <record id="prescription_order_line_state_exception" model="base.kanban.stage">
+        <field name="name">Exception</field>
```
Let's change DUR to Exception in the main module

lasley replied:
It should also be sequenced between Ordered and PV2, not between PV2 and Complete

lasley commented on medical_prescription_state_verify/models/medical_prescription_order.py:
```diff
@@ -18,29 +18,16 @@ class MedicalPrescriptionOrder(models.Model):
             that has been defined in `_ALLOWED_CHANGE_STATES`)
         _ALLOWED_CHANGE_STATES: `list` of keys defining status types in which
             an RX can be moved to after being verified
+            4 - Verified
```
What are these defining? Are they database IDs or something?

kellylougheed replied:
These are the stage_ids, which seem to be set by the sequence of stages in the data files. I wanted to define their states by their name, but I was having difficulty getting the name from the stage_id. Are the ids OK, or should I try to use the names?

kellylougheed replied:
BTW, I am currently working on using the names instead of the numbers

lasley replied:
Damn this strategy isn't going to work anymore without aggregating. Ok let's hold this & swoop back after `T0071 - Create base_kanban_stage_state`

kellylougheed replied:
OK, I'm going to push what I have so far and then hopefully get started on that ticket soon!

kellylougheed replied:
@lasley should I cherry-pick from https://github.com/OCA/server-tools/pull/760 and keep working on this?

lasley commented on medical_prescription_state_verify/models/medical_prescription_order.py:
```diff
@@ -18,29 +18,16 @@ class MedicalPrescriptionOrder(models.Model):
     """
 
-    _inherit = 'medical.prescription.order'
+    _inherit = ['medical.prescription.order', 'base.kanban.abstract']
```
Why inheriting kanban? This should be handled with a dependency to the prescription stage module

lasley commented on medical_prescription_state_verify/models/medical_prescription_order.py:
```diff
@@ -51,33 +38,30 @@ def write(self, vals, ):
         """
-        for rec_id in self:
-            if rec_id.state_type == 'verified':
+        if self.stage_id.name == 'Verified':
```
This will not work. We need an option on the stage to indicate it is a "verified" stage

kellylougheed replied:
OK, I will add a "Verified" property to base_kanban_stage.

lasley replied:
Just in an inherited model, not in that module. Let's call it is_verified

kellylougheed replied:
OK, will do

lasley commented on sale_crm_medical_prescription/tests/test_crm_lead.py:
```diff
@@ -35,8 +35,8 @@ def test_compute_patient_ids(self):
         )
-        self.assertEquals(
-            [patient.id],
+        self.assertTrue(
```
`self.assertIn`

kellylougheed committed:
Fix sale_crm_medical_prescription test

woodbrettm commented on sale_crm_medical_prescription/tests/test_crm_lead.py:
```diff
@@ -24,21 +24,24 @@ def setUp(self):
-        )
+        for id in [patient.id]:
+            self.assertIn(
+                id,
```
you can probably just put `patient.id` here instead of having to put the loop

woodbrettm commented on sale_crm_medical_prescription/tests/test_crm_lead.py:
```diff
@@ -24,21 +24,24 @@ def setUp(self):
-            self.crm_lead_1.prescription_order_ids.ids
-        )
+        rx_orders = [self.rx_order_9.id, self.rx_order_10.id]
+        for rx in rx_orders:
```
Nice yeah this is a much better solution than what I've been doing. Directly comparing the arrays has become problematic.

lasley commented on sale_crm_medical_prescription/tests/test_crm_lead.py:
```diff
@@ -24,21 +24,24 @@ def setUp(self):
-            [self.rx_order_9.id, self.rx_order_10.id],
-            self.crm_lead_1.prescription_order_ids.ids
-        )
+        rx_orders = [self.rx_order_9.id, self.rx_order_10.id]
```
Compare the records, not the IDs

lasley commented on sale_crm_medical_prescription/tests/test_crm_lead.py:
```diff
@@ -24,21 +24,24 @@ def setUp(self):
-            [patient.id],
-            self.crm_lead_1.patient_ids.ids,
-        )
+        for id in [patient.id]:
```
Compare records, not IDs

kellylougheed committed:
Fix tests

kellylougheed commented:
Changes made @bmw95 @lasley
